### PR TITLE
Remove cache

### PR DESCRIPTION
**Pull request by meyskens**
This removes the internal cache system by passing JSON strings as
rendered property to the data resource. This gets passed to
ignition_config instead of the ID.
This uses the Terraform state as cache instead of the old in memory module.

This change is breaking as it will need to have all ID references replaced.

Signed-off-by: Maartje Eyskens <maartje@eyskens.me>

### CODE REVIEW

**Comment by alexsomesan** [comment]
I'm fine merging this as mitigation since there is limited capacity right now to investigate the full chain of events that causes #57.

**Comment by meyskens** [comment]
@alexsomesan thx for the heads up! I will try to finish this PR (mostly tests and cleanup) so we can move on!

**Comment by meyskens** [comment]
I changed all tests to no longer use IDs, ready for review now!

**Comment by meyskens** [comment]
@seh the breaking change is best illustrated in the tests:
https://github.com/terraform-providers/terraform-provider-ignition/pull/56/files#diff-ce99dcb7b0e71397a93e01fe364c3f41R97
Instead of referring to IDs now actual json is passed

Currently waiting on somebody to review this to get it merged.
cc @alexsomesan

**Comment by meyskens** [comment]
Storing it to disk could also be a solution but gave some issues for me in the past. Keeping it as rendered somehow also stores it to disk but in the state file.

And you're right while testing things out I totally forgot the documentation changes!

File: website/docs/d/networkd_unit.html.md [hunk 1]
```diff
@@ -31,4 +31,4 @@ The following arguments are supported:
 
 The following attributes are exported:
 
-* `id` - ID used to reference this resource in _ignition_config_.
\ No newline at end of file
+* `rendered` - The rendered template to reference this resource in _ignition_config_.g_.
```
Comment (meyskens) [inline website/docs/d/networkd_unit.html.md]:
```suggestion
* `rendered` - The rendered template to reference this resource in _ignition_config_.
```

File: website/docs/index.html.markdown [hunk 2]
```diff
@@ -35,14 +35,14 @@ data "ignition_systemd_unit" "example" {
...
+    data.ignition_systemd_unit.example.rendered,
   ]
 }
 
 # Create a CoreOS server using the Igntion config.
 resource "aws_instance" "web" {
   # ...
 
-  user_data = "${data.ignition_config.example.rendered}"
+  user_data = data.ignition_config.example.rendered"
```
Comment (meyskens) [inline website/docs/index.html.markdown]:
```suggestion
  user_data = data.ignition_config.example.rendered
```

**Comment by seh** [comment]
I've since tested this more comprehensively, and it's working well for us.

@alexsomesan, [you mentioned about eight weeks ago](https://github.com/terraform-providers/terraform-provider-ignition/pull/56#issuecomment-509658076) that you'd be willing to merge this. Can we proceed? Without this fix merged and released, it's difficult to migrate to Terraform 0.12. Every member of the team would need to set up a home-built provider in the _~/.terraform.d/plugins_ tree.

**Comment by smalltown** [comment]
test the PR in Terraform 0.12.8, everything works fine like @seh said, @alexsomesan could you help to take a look for this PR and merge it, thanks ^^

**Comment by seh** [comment]
As I've tested this more, I now have a concern: It seems that my "user_data" fields to my EC2 instances and launch configurations change every time that I use _terraform plan_ after _terraform apply_.

I'm going to capture the user data strings between two runs to see if indeed the Ignition provider is now generating something that varies with each run, despite the input not changing.

**Comment by seh** [comment]
Well, comparing the resulting "user_data" values—both the raw value as compact JSON, and then run through _jq_—reveal no differences, yet every time I run _terraform plan_, Terraform says that the user data value is forcing replacement of the "aws_instance" resources. The same goes for my launch configurations.

I'll try to look into this more over the weekend.

**Comment by seh** [comment]
After reading several Terraform GitHub issues on the commute home (search for, say, "Terraform user_data force" and take a deep breath), I discovered that I've had a `depends_on` attribute in my `ignition_file` data sources, expressing a dependency explicitly on the related `aws_s3_bucket_object` resource, with this comment in my code explaining why from about a year and a half ago:

> Work around Terraform defect with data sources depending on resource attributes, in which it fails to update the former when the latter changes.

Removing that `depends_on` attribute today allows _terraform plan_ to relent on finding spurious changes in the "user data" content. What I don't know yet, though, is whether the data source will properly reflect changes in the S3 object content. We're now trying this with a much evolved Terraform version, which I hope has since fixed that defect.

To back up from [my accusation earlier](https://github.com/terraform-providers/terraform-provider-ignition/pull/56#issuecomment-529001757), I don't think this is a problem with this provider. I hope to get in more testing over the weekend to confirm that I was able to work around this spurious change problem.

**Comment by lipsa-vlad** [comment]
I guess this will also fix #55.

**Comment by seh** [comment]
Yes, I think it does fix that one. My configuration uses both, and works with this provider.

Looking at other recent committers here besides @alexsomesan, there's @appilon and @pbrit. Would either of you two be able to help us move forward with this patch?

**Comment by sdlarsen** [comment]
This PR fixes ignition woes for me. Would be nice to see it merged.

**Comment by smalltown** [comment]
Hi @alexsomesan and @appilon 

Sorry to bother you,
This PR has tested by many people,
and it's a blocker when upgrading Terraform from 0.11 to 0.12,
Could you please kindly review and merge it,

Thank for considering my request

**Comment by yokiworks** [comment]
Any progress on this one?

**Comment by seh** [comment]
@alexsomesan, for us to get the full value from this patch, we'll need a fresh release for the provider. What's involved in issuing a release?

**Comment by seh** [comment]
This change is now available in [release 1.2.0](https://github.com/terraform-providers/terraform-provider-ignition/releases/tag/v1.2.0).

**Pull request merged**